core/utils/log.cpp

```diff
@@ -65,47 +67,96 @@ NS_LOCAL
-  FILE* file(iresearch::logger::level_t level) const { return out_[level].file_; }
+  bool enabled(iresearch::logger::level_t level) const { return noop_log_appender != out_[level].appender_.load(std::memory_order_relaxed); }
 
-  logger_ctx& output(iresearch::logger::level_t level, FILE* out) {
-    out_[level].file_ = out ? out : dev_null();
+  logger_ctx& output(iresearch::logger::level_t level, iresearch::logger::log_appender_callback_t appender, void* context) {
+    static std::mutex mutex;
+    SCOPED_LOCK(mutex); // only log level parallel updates are not thread safe. Using log contexts is safe even while concurrent update is ongoing
+    if (appender != nullptr) {
+      auto& ctx = out_[level];
+      ctx.appender_.store(noop_log_appender, std::memory_order_relaxed); // to play safe - as noop never uses context and noop log never breaks
+      ctx.appender_context_.store(context, std::memory_order_release);
+      ctx.appender_.store(appender, std::memory_order_relaxed);
+    } else {
+      out_[level].appender_.store(noop_log_appender, std::memory_order_relaxed);
+    }
     return *this;
   }
 
-  logger_ctx& output_le(iresearch::logger::level_t level, FILE* out) {	
-    for (size_t i = 0, count = IRESEARCH_COUNTOF(out_); i < count; ++i) {	
-      output(static_cast<iresearch::logger::level_t>(i), i > level ? nullptr : out);	
-    }	
+  logger_ctx& output_le(iresearch::logger::level_t level, iresearch::logger::log_appender_callback_t appender, void* context) {
+    for (size_t i = 0, count = IRESEARCH_COUNTOF(out_); i < count; ++i) {
+      output(static_cast<iresearch::logger::level_t>(i), i > level ? nullptr : appender, context);
+    }
     return *this;
   }
 
+  logger_ctx& output(iresearch::logger::level_t level, FILE* out) {
+    return output(level, out != nullptr? fd_log_appender: nullptr, out);
+  }
+
+  logger_ctx& output_le(iresearch::logger::level_t level, FILE* out) {
+    return output_le(level, out != nullptr ? fd_log_appender : nullptr, out);
+  }
+
   irs::logger::level_t stack_trace_level() { return stack_trace_level_; }
   logger_ctx& stack_trace_level(irs::logger::level_t level) {
     stack_trace_level_ = level;
     return *this;
   }
 
+  void log(const char* function, const char* file, int line,
+           iresearch::logger::level_t level, const char* message, size_t len) {
+    const auto& log_ctx = out_[level];
+    auto context = log_ctx.appender_context_.load(std::memory_order_acquire);
+    auto appender = log_ctx.appender_.load(std::memory_order_relaxed);
+    appender(context, function, file, line, level, message, len);
+  }
+
  private:
   struct level_ctx_t {
-    FILE* file_;
-    level_ctx_t(): file_(dev_null()) {}
+    std::atomic<irs::logger::log_appender_callback_t> appender_;
+    std::atomic<void*> appender_context_;
+    level_ctx_t(): appender_(noop_log_appender), appender_context_(nullptr) {}
```

Comment:
let's use 16-bytes structure instead
```suggestion
struct alignas(IRESEARCH_CMPXCHG16B_ALIGNMENT) level_ctx_t {
  void* context{nullptr};
  log_appender_callback_t appender{noop_log_appender};
};
```